plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py

```diff
@@ -0,0 +1,252 @@
+                # TODO: handle this better
+                # _vdi_sanitize(vdi, opq, db, cb)
+                is_snapshot = bool(vdi.volume.snap)
+                if is_snapshot:
+                    path = os.path.basename(sr) + '/'+ str(vdi.volume.id) + '@' + str(vdi.volume.id)
+                else:
+                    path = os.path.basename(sr) + '/'+ str(vdi.volume.id)
+                custom_keys = db.get_vdi_custom_keys(vdi.uuid)
+                vdi_uuid = vdi.uuid
+
+        out = ZFSUtil.get_vsize(dbg, path)
+        if is_snapshot:
+            psize = int(out[0])
+        else:
+            psize = int(out[0]) + int(out[1])
+        vdi_uri = cb.getVolumeUriPrefix(opq) + vdi_uuid
+        return {
+            'uuid': vdi.uuid,
+            'key': vdi.uuid,
+            'name': vdi.name,
+            'description': vdi.description,
+            'read_write': not bool(vdi.volume.snap),
+            'virtual_size': vdi.volume.vsize,
+            'physical_utilisation': psize,
+            'uri': [image_format.uri_prefix + vdi_uri],
+            'keys': custom_keys,
+            'sharable': False
+        }
+
+    def snapshot(self, dbg, sr, key):
+        snap_uuid = str(uuid.uuid4())
+        cb = self.callbacks
+        with VolumeContext(cb, sr, 'w') as opq:
+            result_volume_id = ''
+            with PollLock(opq, 'gl', cb, 0.5):
+                with cb.db_context(opq) as db:
+                    vdi = db.get_vdi_by_id(key)
+                    image_format = ImageFormat.get_format(vdi.image_type)
+                    image_utils = image_format.image_utils
+
+                    vol_id = (vdi.volume.id if vdi.volume.snap == 0 else
+                              vdi.volume.parent_id)
+
+                    vol_path = cb.volumeGetPath(opq, str(vol_id))
+
+                    # snapshot inherits volume's parent
+                    if vdi.volume.parent_id == None:
+                        snap_volume = db.insert_new_volume(vdi.volume.vsize, vdi.image_type)
+                    else:
+                        snap_volume = db.insert_child_volume(vdi.volume.parent_id,
+                                                         vdi.volume.vsize)
+                    db.set_volume_as_snapshot(snap_volume.id)
+                    db.insert_vdi(vdi.name, vdi.description,
+                                  snap_uuid, snap_volume.id, False)
+
+                    db.update_volume_parent(vdi.volume.id, snap_volume.id)
+
+                    result_volume_id = str(snap_volume.id)
+                    ZFSUtil.snapshot(dbg, str(snap_volume.id),
+                                 os.path.basename(sr) + '/'+ str(vdi.volume.id), False)
+                    path = os.path.basename(sr) + '/'+ str(vdi.volume.id) + '@' + str(snap_volume.id)
+                    ZFSUtil.clone(dbg, path, os.path.basename(sr) + '/' + str(snap_volume.id))
+                    ZFSUtil.promote(dbg, os.path.basename(sr) + '/' + str(snap_volume.id))
+        psize = 0
```

Comment:
Why?

Reply:
I added a comment to explain why.

Reply:
I fixed this by using the value from get_vsize which is the used value. However, in zfs this is always 0 for a snapshot since new snapshots not consume space.